### PR TITLE
Metric service

### PR DESCRIPTION
**Pull request by varadeth**

### CODE REVIEW

**Comment by tejash-jl** [comment]
https://github.com/Sunbird-RC/community/issues/380

File: java/registry/src/main/java/dev/sunbirdrc/registry/service/MaskService.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+
+import static dev.sunbirdrc.registry.middleware.util.JSONUtil.convertObjectJsonString;
+
+@Service
+public class MaskService {
```
Comment (tejash-jl) [inline java/registry/src/main/java/dev/sunbirdrc/registry/service/MaskService.java:21]: update class name

File: java/registry/src/main/java/dev/sunbirdrc/registry/config/GenericConfiguration.java [hunk 2]
```diff
@@ -287,6 +287,32 @@ public ISearchService searchService() {
 		return searchProvider.getSearchInstance(searchProviderName, isElasticSearchEnabled());
 	}
 
+	@Bean(name = "kafkaEventService")
+	@ConditionalOnProperty(name = "metrics.providerName", havingValue = "dev.sunbirdrc.registry.service.impl.KafkaEventService", matchIfMissing = true)
```
Comment (tejash-jl) [inline java/registry/src/main/java/dev/sunbirdrc/registry/config/GenericConfiguration.java:291]: add Conditional annotation at class level

File: java/registry/src/main/java/dev/sunbirdrc/registry/service/mask/EmitMap.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package dev.sunbirdrc.registry.service.mask;
+
+import dev.sunbirdrc.registry.model.EventConfig;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class EmitMap {
```
Comment (tejash-jl) [inline java/registry/src/main/java/dev/sunbirdrc/registry/service/mask/EmitMap.java:8]: use factory pattern

File: java/registry/src/main/java/dev/sunbirdrc/registry/service/mask/MaskEmitStrategy.java [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package dev.sunbirdrc.registry.service.mask;
+
+import org.apache.commons.lang3.StringUtils;
+
+public class MaskEmitStrategy implements IEmitStrategy {
+    @Override
+    public String updateValue(String value) {
+        String replacement = StringUtils.repeat('X', value.length()/2);
```
Comment (tejash-jl) [inline java/registry/src/main/java/dev/sunbirdrc/registry/service/mask/MaskEmitStrategy.java:8]: check for empty or null

File: services/metrics/cron/cron.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	}
+	cron.redisClient = redis.NewClient(redisOptions)
+	locker, err := redislock.NewRedisLocker(cron.redisClient, redislock.WithTries(1))
+	if err != nil {
+		log.Errorf("Failed Gaining Lock Because : %v", err)
+	}
+
+	cronJob := gocron.NewScheduler(time.UTC)
+	cronJob.WithDistributedLocker(locker)
+	_, err = cronJob.Every(7).Day().At("19:30").Do(cron.SaveWeeklyMetrics, database)
```
Comment (tejash-jl) [inline services/metrics/cron/cron.go:39]: date and time be configurable

File: services/metrics/cron/cron.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+	if err != nil {
+		log.Errorf("Failed Creating a scheduler : %v", err)
+	}
+	CronObj = &cron
+}
+
+func (cron *Cron) SaveWeeklyMetrics(db models.IDatabase) {
```
Comment (tejash-jl) [inline services/metrics/cron/cron.go:46]: see if this parameter is required

File: services/metrics/cron/cron.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	days := 7
```
Comment (tejash-jl) [inline services/metrics/cron/cron.go:47]: use the config

File: services/metrics/cron/cron.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	backDatedDate := time.Now().Add(-1 * time.Duration(24*time.Duration(days)*time.Hour)).Format("2006-01-02")
+	whereClause := "WHERE date > '" + backDatedDate + "'"
+	aggregate := db.GetAggregates(whereClause)
+	bytes, err := json.Marshal(aggregate)
+	if err != nil {
+		log.Infof("Error in marshalling : %v", err)
+	}
+	ctx := context.Background()
+	log.Infof(string(bytes))
+	err = cron.redisClient.Set(ctx, "weeklyUpdates", string(bytes), 24*time.Duration(days)*time.Hour).Err()
```
Comment (tejash-jl) [inline services/metrics/cron/cron.go:57]: set the value in redis without expiry

File: services/metrics/interfaces/metrics-api.yaml [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+        - application/json
+      produces:
+        - application/json
+      security: []
+      responses:
+        '200':
+          description: OK
+          schema:
+            $ref: '#/definitions/LoginResponse'
+  /v1/aggregates/weekly:
```
Comment (tejash-jl) [inline services/metrics/interfaces/metrics-api.yaml:28]: without weekly

File: services/metrics/models/Clickhouse.go [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+		log.Fatal(err)
+	}
+}
+
+func createTableIfNotExists(tableName string, c *Clickhouse) error {
+	ctx := context.Background()
+	createTableCmd := `CREATE TABLE IF NOT EXISTS ` + strings.ToLower(tableName) + ` (
+		operationType 	String,
+		entity			JSON,
+		date 			Date,
```
Comment (tejash-jl) [inline services/metrics/models/Clickhouse.go:41]: createdAt

File: java/registry/src/main/java/dev/sunbirdrc/registry/service/impl/RegistryServiceImpl.java [hunk 11]
```diff
@@ -160,8 +167,13 @@ public Vertex deleteEntityById(Shard shard, String userId, String uuid) throws E
                     if (isElasticSearchEnabled()) {
                         callESActors(null, "DELETE", index, uuid, tx);
                     }
-
-
+                    JsonNode deletedNode = vertexReader.constructObject(vertex);
```
[the remark this answers is not in the document]
Comment (varadeth) [reply java/registry/src/main/java/dev/sunbirdrc/registry/service/impl/RegistryServiceImpl.java:170]: this is read one time only. it is just that vertex is converted to JsonNode here

File: java/registry/src/main/java/dev/sunbirdrc/registry/model/event/Event.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import lombok.Builder;
+
+@JsonSerialize
+@Builder
+public class Event {
+    @JsonProperty("eid")
```
[the remark this answers is not in the document]
Comment (varadeth) [reply java/registry/src/main/java/dev/sunbirdrc/registry/model/event/Event.java:11]: if this is used, then only TelemetrObject's object is getting sent
```example
{
  "object" : {
    "id" : "{{someOsid}}",
    "type" : "{{schemaName}}"
  }
}
```

File: docker-compose.yml [hunk 13]
```diff
@@ -39,6 +39,9 @@ services:
       - connectionInfo_uri=jdbc:postgresql://db:5432/registry
       - connectionInfo_username=postgres
       - connectionInfo_password=postgres
+      - event_enabled=false
+      - event_topic=metrics
```
Comment (tejash-jl) [inline docker-compose.yml:43]: need to change the topic

**Review by tejash-jl — APPROVED** [review]

**Pull request merged**